tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,255 @@
+      check topic notin currentGossip.topics
+      if topic in currentGossip.mesh:
+        check currentGossip.mesh[topic].len == 0
+      else:
+        check topic notin currentGossip.mesh
+      if topic in currentGossip.gossipsub:
+        check currentGossip.gossipsub[topic].len == 0
+      else:
+        check topic notin currentGossip.gossipsub
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  # Test subscribing and unsubscribing multiple topics
+  asyncTest "handle SUBSCRIBE and UNSUBSCRIBE multiple topics":
+    # Given 3 nodes and multiple topics
+    let
+      numberOfNodes = 3
+      topics = ["topic1", "topic2", "topic3"].toSeq()
+      nodes = generateNodes(numberOfNodes, gossip = true)
+
+    await allFuturesThrowing(nodes.mapIt(it.switch.start()))
+
+    # When nodes subscribe and then unsubscribe to multiple topics
+    await subscribeNodes(nodes)
+    for node in nodes:
+      for topic in topics:
+        node.subscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # Then all nodes should be subscribed to the topics initially
+    for node in nodes:
+      let currentGossip = GossipSub(node)
+      check currentGossip.topics.len == topics.len
+      for topic in topics:
+        check currentGossip.gossipsub[topic].len == numberOfNodes - 1
+
+    # When they unsubscribe from all topics
+    for node in nodes:
+      for topic in topics:
+        node.unsubscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # Then topics should be removed from mesh and gossipsub
+    for node in nodes:
+      let currentGossip = GossipSub(node)
+      for topic in topics:
+        check topic notin currentGossip.topics
+        check topic notin currentGossip.mesh
+        check topic notin currentGossip.gossipsub
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  # Test ensuring the number of subscriptions does not exceed a set limit
+  asyncTest "subscription limit test":
+    # Given one node and a subscription limit of 10 topics
+    let
+      topicCount = 15
+      gossipSubParams = 10
+      topicNames = toSeq(mapIt(0 .. topicCount - 1, "topic" & $it))
+      numberOfNodes = 1
+      nodes = generateNodes(numberOfNodes, gossip = true)
+
+    await allFuturesThrowing(nodes.mapIt(it.switch.start()))
+
+    # When attempting to subscribe to 15 topics
+    let gossipSub = GossipSub(nodes[0])
+    gossipSub.topicsHigh = gossipSubParams
+
+    for topic in topicNames:
+      if gossipSub.topics.len < gossipSub.topicsHigh:
+        gossipSub.subscribe(
+          topic,
+          proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+            discard
+          ,
+        )
+      else:
+        # Then the subscription count should not exceed the limit
+        check gossipSub.topics.len == gossipSub.topicsHigh
+
+    check gossipSub.topics.len == gossipSub.topicsHigh
```

Comment:
I would check the `gossipsub.topics.len` against `gossipSubParams`, which is the limit you defined here. That would make the check accurate.

Also, maybe it'd be interesting to connect the nodes and check for `mesh` and `gossipsub` population.